db/src/db.rs

```diff
@@ -401,56 +409,60 @@ impl TypedSQLValue for TypedValue {
 }
 
 /// Read the partition map materialized view from the given SQL store.
-pub fn read_partition_map(conn: &rusqlite::Connection) -> Result<PartitionMap> {
+fn read_partition_map(conn: &rusqlite::Connection) -> Result<PartitionMap> {
     let mut stmt: rusqlite::Statement = conn.prepare("SELECT part, start, idx FROM parts")?;
     let m = stmt.query_and_then(&[], |row| -> Result<(String, Partition)> {
         Ok((row.get_checked(0)?, Partition::new(row.get_checked(1)?, row.get_checked(2)?)))
     })?.collect();
     m
 }
 
-/// Read the schema materialized view from the given SQL store.
-pub fn read_schema(conn: &rusqlite::Connection, ident_map: &IdentMap) -> Result<Schema> {
-    let mut stmt: rusqlite::Statement = conn.prepare("SELECT ident, attr, value, value_type_tag FROM schema")?;
-    let r: Result<Vec<(symbols::NamespacedKeyword, symbols::NamespacedKeyword, TypedValue)>> = stmt.query_and_then(&[], |row| {
-        // Each row looks like :db/index|:db/valueType|28|0.  Observe that 28|0 represents a
-        // :db.type/ref to entid 28, which needs to be converted to a TypedValue.
-        // TODO: don't use textual ident and attr; just use entids directly.
-        let symbolic_ident: String = row.get_checked(0)?;
-        let symbolic_attr: String = row.get_checked(1)?;
-        let v: rusqlite::types::Value = row.get_checked(2)?;
-        let value_type_tag: i32 = row.get_checked(3)?;
-        let typed_value = TypedValue::from_sql_value_pair(v, value_type_tag)?;
-
-        let ident = to_namespaced_keyword(&symbolic_ident);
-        let attr = to_namespaced_keyword(&symbolic_attr);
-        match (ident, attr, typed_value) {
-            (Ok(ident), Ok(attr), typed_value) => Ok((ident, attr, typed_value)),
-            (Err(e), _, _) => Err(e),
-            (_, Err(e), _) => Err(e),
+/// Read the ident map materialized view from the given SQL store.
+fn read_ident_map(conn: &rusqlite::Connection) -> Result<IdentMap> {
+    let v = read_materialized_view(conn, "idents")?;
+    v.into_iter().map(|(e, a, typed_value)| {
+        if a != entids::DB_IDENT {
+            bail!(ErrorKind::NotYetImplemented(format!("bad idents materialized view: expected :db/ident but got {}", a)));
         }
-    })?.collect();
+        if let TypedValue::Keyword(keyword) = typed_value {
+            Ok((keyword, e))
+        } else {
+            bail!(ErrorKind::NotYetImplemented(format!("bad idents materialized view: expected [entid :db/ident keyword] but got [entid :db/ident {:?}]", typed_value)));
+        }
+    }).collect()
+}
 
-    r.and_then(|triples| Schema::from_ident_map_and_triples(ident_map.clone(), triples))
+/// Read the schema materialized view from the given SQL store.
+fn read_schema_map(conn: &rusqlite::Connection) -> Result<SchemaMap> {
+    let entid_triples = read_materialized_view(conn, "schema")?;
+    let mut schema_map = SchemaMap::default();
+    metadata::update_schema_map_from_entid_triples(&mut schema_map, entid_triples)?;
```

Comment:
Can you push this into a method on `SchemaMap`? `SchemaMap::with_entid_triples`?